File: python/stack/stack.py
from typing import List, Iterator


class Stack[T]:
    def __init__(self):
        self._items: List[T] = []

    def size(self) -> int:
        return len(self._items)

    def is_empty(self) -> bool:
        return len(self._items) == 0

    def push(self, item: T):
        self._items.append(item)

    def pop(self) -> T:
        if not self._items:
            raise IndexError("pop from empty stack")
        return self._items.pop()

    def peek(self) -> T:
        if not self._items:
            raise IndexError("peek from empty stack")
        return self._items[-1]

    def __iter__(self) -> Iterator[T]:
        for item in reversed(self._items):
            yield item

File: rust/stack/src/stack.rs
pub struct Stack<T: Clone> {
    elements: Vec<T>,
}

impl<T: Clone> Stack<T> {
    pub fn new() -> Self {
        Stack {
            elements: Vec::new(),
        }
    }

    pub fn size(&self) -> usize {
        self.elements.len()
    }

    pub fn push(&mut self, item: T) {
        self.elements.push(item);
    }

    pub fn pop(&mut self) -> Option<T> {
        self.elements.pop()
    }

    pub fn peek(&self) -> Option<T> {
        if self.elements.len() == 0 {
            return None;
        }
        Some(self.elements[self.elements.len() - 1].clone())
    }

    pub fn is_empty(&self) -> bool {
        self.elements.is_empty()
    }
}
